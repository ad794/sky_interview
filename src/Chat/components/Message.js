import React, { Component } from 'react';
import { View, Text, TouchableWithoutFeedback } from 'react-native';
import PropTypes from 'prop-types';
import Timestamp from './Timestamp';
import styles from './styles';

class Message extends Component {
  static propTypes = {
    message: PropTypes.string,
    timestamp: PropTypes.string,
    member: PropTypes.object
  };

  constructor() {
    super();
    this.state = {
      showEmail: false
    };
  }

  onPressInHandler = () => {
    this.setState({
      showEmail: true
    });
  };
  onPressOutHandler = () => {
    this.setState({
      showEmail: false
    });
  };

  render() {
    const { message, member, timestamp } = this.props;
    return (
      <View style={styles.wrapper}>
        <View style={styles.message}>
          <TouchableWithoutFeedback
            testID="showEmail"
            onPressIn={this.onPressInHandler}
            delayPressOut={1200}
            onPressOut={this.onPressOutHandler}
          >
            <View>
              <Text testID="messageText" style={styles.message}>
                {message}
              </Text>
              {this.state.showEmail && (
                <Text testID="emailText" style={styles.email}>
                  {member && member.email}
                </Text>
              )}
            </View>
          </TouchableWithoutFeedback>
        </View>
        <Timestamp timestamp={timestamp} />
      </View>
    );
  }
}

export default Message;
