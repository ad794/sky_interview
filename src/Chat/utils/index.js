export const timestampToDate = (timestamp) => {
  const date = new Date(timestamp * 1000);
  return `${date.toLocaleDateString()} at ${formatTime(date)}`;
};

formatTime = (date_obj) => {
  // formats a javascript Date object into a 12h AM/PM time string
  let hour = date_obj.getHours();
  let minute = date_obj.getMinutes();
  const amPM = hour > 11 ? 'PM' : 'AM';
  if (hour > 12) {
    hour -= 12;
  } else if (hour == 0) {
    hour = '12';
  }
  if (minute < 10) {
    minute = `0${minute}`;
  }
  return `${hour}:${minute}${amPM}`;
};
export const findById = function (tree, testID) {
  if (tree.props && tree.props.testID === testID) {
    return tree;
  }
  if (tree.children && tree.children.length > 0) {
    const childs = tree.children;
    for (let i = 0; i < childs.length; i++) {
      const item = findById(childs[i], testID);
      if (typeof item !== 'undefined') {
        return item;
      }
    }
  }
};
