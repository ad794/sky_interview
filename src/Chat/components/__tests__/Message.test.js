import React from 'react';
import renderer from 'react-test-renderer';
import Message from '../Message';

import { findById } from '../../utils';

describe('message component', () => {
  const mock = {
    message: 'mockdata',
    timestamp: '2016-11-09T05:04:58Z',
    member: { email: 'test@test.com' }
  };
  const messageComponentTree = renderer.create(<Message {...mock} />).toJSON();
  const messageComponent = renderer.create(<Message {...mock} />).getInstance();
  test('renders correctly', () => {
    expect(messageComponentTree).toMatchSnapshot();
  });

  it('renders with correct props', () => {
    expect(messageComponent.props).toEqual(mock);
  });

  it('renders and displays correct message text ', () => {
    const element = findById(messageComponentTree, 'messageText');
    expect(element).toBeDefined();
    expect(element.children[0]).toBe(mock.message);
  });

  it('set correct state on onPressIn and onPressOut handlers', () => {
    messageComponent.onPressInHandler();
    expect(messageComponent.state.showEmail).toEqual(true);
    messageComponent.onPressOutHandler();
    expect(messageComponent.state.showEmail).toEqual(false);
  });
});
