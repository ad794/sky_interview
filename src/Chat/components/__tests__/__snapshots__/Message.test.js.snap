// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message component renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "marginLeft": 10,
      }
    }
  >
    <View
      accessible={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="showEmail"
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "flex": 1,
            "justifyContent": "center",
            "marginLeft": 10,
          }
        }
        testID="messageText"
      >
        mockdata
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flex": 1,
        "justifyContent": "flex-end",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      1970-1-1 at 12:00AM
    </Text>
  </View>
</View>
`;
